pkg/rancher/rancher.go

```diff
@@ -396,11 +399,29 @@ func (r *Rancher) ListenAndServe(ctx context.Context) error {
 		go func() {
 			logrus.Infof("Waiting for %s imperative API to be ready", r.aggregationRegistrationTimeout)
 
-			select {
-			case <-r.kubeAggregationReadyChan:
-				logrus.Info("kube-apiserver connected to imperative api")
-			case <-time.After(r.aggregationRegistrationTimeout):
-				logrus.Fatal("kube-apiserver did not contact the rancher imperative api in time, please ensure k8s is configured to support api extension")
+			ctx, cancel := context.WithTimeout(context.Background(), r.aggregationRegistrationTimeout)
+			defer cancel()
+
+			apiserviceClient := r.Wrangler.API.APIService()
+			for {
+				select {
+				// Case 1: ExtensionServer was contacted by the Kube API,
+				case <-r.kubeAggregationReadyChan:
+					logrus.Info("kube-apiserver connected to imperative api")
+					ext.SetAggregationCheck(apiserviceClient, true)
+					return
+				// Case 2: ExtensionServer in a different replica was contacted by the Kube API
+				//         Only valid if Steve was configured to wait. Otherwise, it would short-circuit here directly here
+				case <-time.After(5 * time.Second):
+					if r.Steve.WaitForExtensionAPIServer && ext.AggregationPreCheck(apiserviceClient) {
+						logrus.Info("kube-apiserver connected to imperative api")
+						return
+					}
+				// Case 3: the Kube API didn't contact the Extension Server within the configured timeout, interpreted as API Aggregation not being supported in the cluster
+				case <-ctx.Done():
+					ext.SetAggregationCheck(apiserviceClient, false)
+					logrus.Fatal("kube-apiserver did not contact the rancher imperative api in time, please ensure k8s is configured to support api extension")
+				}
```

Comment:
At this point we should consider using [`wait.PollUntilContextTimeout`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollUntilContextTimeout) if we are now manually managing a timeout, an interval, and a condition.